Task/doc updates

rhornung67 opened the pull request:
# Summary

This PR contains a variety of Dev Guide updates/improvements, including:

- Improved release process section
- Update Gitlab CI description to reflect new RADIUSS Shared CI usage
- Improvements to CI maintenance section

Read The Docs content for this PR branch is here: https://raja.readthedocs.io/en/task-doc-updates/sphinx/dev_guide/index.html

MrBurmark commented on docs/sphinx/dev_guide/build_configurations.rst:
```diff
@@ -100,7 +126,7 @@ This section describes the host-config file generation process for RAJA.
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Compiler configurations for Livermore computer platforms are contained
+Compiler configurations for Livermore machines  are contained
```
```suggestion
Compiler configurations for Livermore machines are contained
```

MrBurmark commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,250 @@
+  * **GitLab** instances in the Collaboration Zone (CZ) of the Livermore 
+    Computing (LC) Center run builds and tests in LC platform and compiler 
+    environments important to many RAJA user applications. Execution of RAJA CI 
+    pipelines on LC GitLab resources has restrictions, which are  described 
```
```suggestion
    pipelines on LC GitLab resources has restrictions, which are described 
```

MrBurmark commented on docs/sphinx/dev_guide/build_configurations.rst:
```diff
@@ -39,34 +40,54 @@ They are described briefly in the following sections.
-passed to CMake RAJA tests, example codes, etc.  For example,
+            * create a build space directory in the top-level RAJA directory
+              named for the specified build configuration
+            * load an appropriate version CMake module for the configuration
```
awkward

rhornung67 replied:
Fixed

MrBurmark commented on docs/sphinx/dev_guide/build_configurations.rst:
```diff
@@ -110,50 +136,64 @@ directory:
-See :ref:`ci-label` for more information.
+Running a command like these will generate a host-config file in the top-level 
+RAJA directory whose name contains the machine and OS to which it applies, as 
+well the compiler and version. For example, running the first example command 
```
```suggestion
well as the compiler and version. For example, running the first example command 
```

rhornung67 replied:
Fixed

MrBurmark commented on docs/sphinx/dev_guide/build_configurations.rst:
```diff
@@ -110,50 +136,64 @@ directory:
+Running a command like these will generate a host-config file in the top-level 
+RAJA directory whose name contains the machine and OS to which it applies, as 
+well the compiler and version. For example, running the first example command 
+above will on the LC machine quartz will generate a host-config file named 
```
```suggestion
above on the LC machine quartz will generate a host-config file named 
```

rhornung67 replied:
Fixed

MrBurmark commented on docs/sphinx/dev_guide/ci_tasks.rst:
```diff
@@ -12,99 +12,134 @@
-be adjusted in the build spec information in the associated 
-``<resource>-jobs.yml`` file discussed in the previous section. 
+               If the desired entry is not there, but exists in a newer version 
+               of the RADIUSS Spack Configs project, update the RAJA submitted 
```
```suggestion
               of the RADIUSS Spack Configs project, update the RAJA submodule 
```

rhornung67 replied:
Fixed

rhornung67 commented:
@adrienbernede I would it appreciate it if you would look this over when you have time. I want to make sure the way I've described how the project GitLab CI is coordinated with RADIUSS projects is correct. Thank you.

adrienbernede commented:
@rhornung67 Will do

rhornung67 commented:
@LLNL/raja-core @adrienbernede anyone else want to look this over before it is merged?

adrienbernede commented:
@rhornung67 my bad, if you can wait I'll have a look tomorrow.

rhornung67 commented:
No rush. Thank you.

adrienbernede reviewed (commented):
Review Part 1.

rhornung67 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -12,50 +12,99 @@
+               Fortunately, the vast majority of those changes will have been
+               reviewed, approved, and merged into the develop branch. 
+   
+               To make it easier for team members to review the release 
+               candidate PR, it usually helps to make a companion draft PR 
+               to merge the release candidate branch into develop. This PR
+               will only show the changes to merge into main for the release.
+               Cross-reference the two PRs in their descriptions, tell 
+               reviewers to review the draft PR into develop, but approve 
+               the real PR to merge into main.
```
Good suggestion. I will incorporate your changes.

rhornung67 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -12,50 +12,99 @@
+     ensure that all changes done to finalize the release are included 
+     in the develop branch and future work on that branch.
 
 After a RAJA release is done, there a other tasks that typically need to be 
 performed to update content in other projects. These task are described in
```
Fixed.

rhornung67 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -68,36 +117,37 @@ Release Candidate Branch
-**rc-<release #>.** Please see :ref:`release-label` for a description of how 
-a release candidate branch is used in the release process. 
-
-.. important:: Creating a release candidate branch starts the next release 
-               cycle whereby new work being performed on feature branches can 
-               be merged into the develop branch.
+from the develop branch. Typically, a release candidate branch is named 
+**rc-<release #>**, or similar. Please see :ref:`release-label` for a 
+description of how a release candidate branch is used in the release process. 
```
Good point. I will move.

rhornung67 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -112,16 +162,13 @@ Hotfix Branch
-will address the issue be merged into both develop and main branches.
+will address the issue in both the develop and main branches.
+
+A hotfix branch treated like a release candidate branch and it is used to 
```
Thanks. Fixed.

rhornung67 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -112,16 +162,13 @@ Hotfix Branch
-the key steps for performing a hotfix release are:
+For completeness, the key steps for performing a hotfix (patch) release are:
 
   #. Make a **hotfix** branch from main for a release (hotfix/<issue>), fix the
```
Yup. More clear.

rhornung67 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -157,21 +204,114 @@ After a RAJA release is complete, other tasks are performed to update content
+After each RAJA release, we update the **RAJA Spack Package** and make a PR to
+push it upstream to the `Spack project <https://github.com/spack/spack>`_. The
+Spack package is used in RAJA GitLab CI testing and also RAJA users who use
+Spack to manage their third party library installations. We try maintain the
```
Fixed.

rhornung67 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -157,21 +204,114 @@ After a RAJA release is complete, other tasks are performed to update content
+Spack package is used in RAJA GitLab CI testing and also RAJA users who use
+Spack to manage their third party library installations. We try maintain the
+RAJA Spack package in the Spack project to be as close as possible to the
+one in the RAJA repository, which has a few modifications for our CI testing.
```
I agree with that goal 100%. I'm not sure we're quite there yet. I will leave this as is for now and change when we are confident.

rhornung67 commented:
@adrienbernede I addressed all your review comments about the release process. Thank you. They were all good.

I'm mostly interested in your feedback on the GitLab CI description and maintenance sections. Please look those over when you have time. Thanks again.

adrienbernede commented:
@rhornung67 I’m getting there, will do tomorrow.

rhornung67 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -157,21 +204,114 @@ After a RAJA release is complete, other tasks are performed to update content
+
+The RAJA Spack package is in the file 
+``RAJA/scripts/spack_packages/raja/package.py``. The package is a Python 
+class. The following list contains a description of items to update.
```
I fixed the location of the package in my last commit. I will try to work your other points into the GitLab CI discussion.

rhornung67 replied:
I've addressed your comments here. 

I think your note belongs in the radiuss shared CI docs.

rhornung67 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -157,21 +204,114 @@ After a RAJA release is complete, other tasks are performed to update content
+      if "camp" in self.spec:
+         entries.append(cmake_cache_path("camp_DIR", spec["camp"].prefix)) 
+
+    respectively.
```
Good point. I will add that.

rhornung67 replied:
I added text and an example about conflicts.

adrienbernede commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,252 @@
 
-The tools automatically run RAJA builds and tests when a PR is created and 
-when changes are pushed to a PR branch.
+The tools integrate with the RAJA GitHub project and automatically run RAJA 
+builds and tests when a PR is created and when changes are pushed to a PR 
+branch.
```
Would it be worth adding a "note" stating that the GitLab CI synchronization, although automated, are asynchronous, which may cause delays?

rhornung67 replied:
I don't think so. If there are any delays, they pale in comparison to jobs waiting in queues to start running.

If you think it's important to note, it probably should be noted in the RADIUSS shared CI documentation.

adrienbernede replied:
You are making a point.

Just a remark: While the mirroring is pending the GitLab checks will not appear in the list on GitHub, so you might have a PR seemingly passing all the checks because it is in fact waiting for the GitLab mirroring.

This is not crucial for RAJA because passing the GitLab check is required to merge.

adrienbernede commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,252 @@
-    pipelines on LC Gitlab resources has restrictions described below. If 
-    you have access to LC resources, you can access additional information about
+  * **GitLab** instances in the Collaboration Zone (CZ) of the Livermore 
+    Computing (LC) Center run builds and tests in LC platform and compiler 
+    environments important to many RAJA user applications. Execution of RAJA CI 
```
It took me several tries to read the sentence the intended way, that is:
"run builds and test in (LC (platform & compiler)) environments."
I was reading:
"run builds and test in (LC platform) & (compiler environments)"
which sounded weird.
How about:
```suggestion
    Computing (LC) Center run builds and tests in environments combining LC platforms
    and compilers important to many RAJA user applications. Execution of RAJA CI
```

rhornung67 replied:
I changed it to "Computing (LC) Center run builds and tests on LC HPC platforms using software stacks (compilers, etc.) important to many RAJA user applications."

Is that more clear?

adrienbernede replied:
Very clear.

adrienbernede commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,252 @@
-The Gitlab CI instance used by the RAJA project lives in the Livermore 
+The GitLab CI instance used by the RAJA project lives in the Livermore 
 Computing (LC) Collaboration Zone (CZ). It runs builds and tests in LC 
-resource and compiler environments important to RAJA user applications at LLNL.
+platform and compiler environments important to RAJA user applications at LLNL.
```
Again here,
```suggestion
Computing (LC) Collaboration Zone (CZ). It runs builds and tests in environments
combining LC platforms and compilers important to RAJA user applications at LLNL.
```

rhornung67 replied:
I reworded this similar to what I posted above.

adrienbernede commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,252 @@
+enabled on their GitHub accounts. When these requirements are satisfied, 
+mirroring of a GitHub repo and triggering GitLab CI functionality from GitHub
+can be done. Otherwise, LC GitLab CI checks will not be run for a project. 
+For a compliant LLNL GitHub project, such as RAJA, auto-mirroring of the 
+GitHub repo on LC GitLab is done when changes are pushed to PR branches in the 
+RAJA GitHub project. If you have access to LC platforms, you can learn more 
+about `LC GitLab mirroring <https://lc.llnl.gov/confluence/pages/viewpage.action?pageId=662832265>`_.
```
```suggestion
For a compliant LLNL GitHub project, such as RAJA, auto-mirroring of the 
GitHub repo on LC GitLab is done every 30 minutes at best, triggering tests
on new changes in RAJA GitHub project. If you have access to LC platforms,
you can learn more about `LC GitLab mirroring <https://lc.llnl.gov/confluence/pages/viewpage.action?pageId=662832265>`_.
```

adrienbernede replied:
I am actually investigating the 30 minute rule mentioned here. According to GitLab documentation, a change on GitHub could actually trigger a pipeline on GitLab... if the webhooks are set. And it seems they often are not, but I can’t check by myself on RAJA.

rhornung67 replied:
We don't have webhooks set, but we do have GitLab and Azure checks required to pass before a PR can be merged. So you're saying that if we set webhooks for GitLab, it may start running sooner after changes are pushed to GitHub? If so, I'll look into that.

adrienbernede commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,252 @@
+
+**GitLab CI will not run for a PR branch on a fork of the RAJA repo.** We 
+manually manage contributions made on a fork of the RAJA repo using the 
+procedure described in :ref:`contributing-label`. 
+.. _gitlab_ci_workflow-label:
```
```suggestion
procedure described in :ref:`contributing-label`.

.. _gitlab_ci_workflow-label:
```

rhornung67 replied:
blank line added. Thanks for catching that.

adrienbernede commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,252 @@
-  #. Gitlab launches CI test pipelines. While running, the execution and 
-     pass/fail status may be viewed and monitored in the Gitlab CI GUI.
-  #. For each resource and compiler combination,
+  #. A *mirror* of the RAJA GitHub repo in the RAJA LC CZ GitLab project is 
+     updated whenever the RAJA ``develop`` or ``main`` branches are changed 
+     as well as when any PR branch in the RAJA GitHub project is changed. 
```
Again, under investigation.

rhornung67 replied:
I reworded this to align with your comment.

rhornung67 commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,252 @@
      space.  Then, RAJA and its tests are compiled.
   #. Next, the RAJA tests are run.
-  #. When a test pipeline completes, final results are reported in Gitlab.
-
```
Done.

adrienbernede replied:
@rhornung67 it looks like you replaced the last step. However your schema shows step 6 as being "job reports to Gitlab", but I was suggesting _adding_ a step (7) "Gitlab reports to Github".
In a nutshell: schema and items have diverged.

rhornung67 replied:
Got it. Thanks.

rhornung67 commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,252 @@
+     pass/fail status may be viewed and monitored in the GitLab CI GUI
+     or in the RAJA GitHub project checks section for a PR.
+  #. For each platform and compiler combination,
      `Spack <https://github.com/spack/spack>`_ is used to generate a build 
      configuration in the form of a CMake cache file, or *host-config* file.
```
Changed.

adrienbernede commented on docs/sphinx/dev_guide/ci_tasks.rst:
```diff
@@ -12,99 +12,142 @@
-try updating the submodule to a newer version. If the information
-is still not available, create a branch in the 
-`RADIUSS Spack Configs <https://github.com/LLNL/radiuss-spack-configs>`_ repo, add the needed spec info, and create a pull request.
+A spec containing the compiler and version information for the machine must 
+exist in the `RADIUSS Spack Configs project <https://github.com/LLNL/radiuss-spack-configs>`_. If it does not exist there, create a branch, add the spec entry,
+and create a pull request. After the PR is merged, update the RADIUSS Spack
+Configs submodule in RAJA to be able to run it in GitLab CI.
```
I find it confusing to use `spec` to talk about compilers in spack although it
correct to say that we need the compiler spec to be defined.

```suggestion
.. important:: The build toolchain (compilers, cuda, hip) information for the
               machine must be defined in the appropriate spack configuration files
               (`compilers.yaml` or `packages.yaml`) shared in the
               `RADIUSS Spack Configs project <https://github.com/LLNL/radiuss-spack-configs>`_.
               If it does not exist there, create a branch, add the compiler entry,
               and create a pull request. After the PR is merged, update the RADIUSS
               Spack Configs submodule in RAJA to be able to invoke the new toolchain
               in GitLab CI.
```

adrienbernede commented on docs/sphinx/dev_guide/ci_tasks.rst:
```diff
@@ -12,99 +12,142 @@
+Modifying a configuration
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To change an existing configuration, change the relevant information in the 
+configuration, such variant(s) or target compute architecture, in the 
+appropriate 
+``RAJA/.gitlab/<MACHINE>-build-and-test-extra.yml`` file. Make sure to 
+also modify the job label, so it is descriptive of the configuration.
```
```suggestion
To change a RAJA specific configuration, change the relevant information
in the configuration, such as variant(s) or target compute architecture, in
the appropriate ``RAJA/.gitlab/<MACHINE>-build-and-test-extra.yml``
file. In that case, make sure to also modify the job label, so it is descriptive
of the configuration.
```

adrienbernede commented on docs/sphinx/dev_guide/ci_tasks.rst:
```diff
@@ -12,99 +12,142 @@
+`RADIUSS Shared CI <https://github.com/LLNL/radiuss-shared-ci>`_ project.
+Create a branch there, modify the job entry, and create a pull request.
+
+.. important:: Build spec information used in RAJA GitLab CI pipelines
                must exist in the ``compilers.yaml`` file and/or 
                ``packages.yaml`` file for the appropriate system type in
                the `RADIUSS Spack Configs <https://github.com/LLNL/radiuss-spack-configs>`_ repo.
```
This is a repetition of the final note of previous section.

```suggestion
```

rhornung67 replied:
Yup. Thanks. Fixed.

adrienbernede commented on docs/sphinx/dev_guide/ci_tasks.rst:
```diff
@@ -12,99 +12,142 @@
-run than the default allotted time. In this case, the default time can
-be adjusted in the build spec information in the associated 
-``<resource>-jobs.yml`` file discussed in the previous section. 
+               If the desired entry is not there, but exists in a newer version 
+               of the RADIUSS Spack Configs project, update the RAJA submodule 
+               to use the newer version. If the information does not exist in 
+               any version of the RADIUSS Spack Configs project, create a 
+               branch there, add the needed spec info, and create a pull 
+               request. Then, when that PR is merged, update the RAJA submodule
+               to the new version.
```
Remove the rest of the repetition.

```suggestion
```

rhornung67 replied:
I made different changes that should have the same resulting effect.

adrienbernede commented on docs/sphinx/dev_guide/ci_tasks.rst:
```diff
@@ -12,99 +12,142 @@
+Changing run parameters
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The parameters for each system/system-type on which we run GitLab CI for
```
```suggestion
The parameters for each system/scheduler on which we run GitLab CI for
```

rhornung67 replied:
changed.

adrienbernede commented on docs/sphinx/dev_guide/ci_tasks.rst:
```diff
@@ -12,99 +12,142 @@
+    RUBY_BUILD_AND_TEST_JOB_ALLOC: "--time=60 --nodes=1"
+  extends: .build_and_test_on_ruby
 
+This example sets the build and test allocation time to 60 minutes and to 
+run on one node. 
```
```suggestion
This example sets the build and test allocation time to 60 minutes and the
run resource to one node. 
```

rhornung67 replied:
fixed

adrienbernede commented on docs/sphinx/dev_guide/ci_tasks.rst:
```diff
@@ -12,99 +12,142 @@
+      SPEC: " ${PROJECT_LASSEN_VARIANTS} %clang@ibm.9.0.0 ${PROJECT_LASSEN_DEPS}"
+    extends: .build_and_test_on_lassen
+    allow_failure: true
 
```
Here we should add an "important note" about overriding specs. This was done well in the ci.rst:
Something like:

```suggestion

.. important:: When a shared job needs to be modified for RAJA specifically, we call
             that "overriding": The job label must be kept the same as in the 
             ``<MACHINE>-build-and-test.yml`` file in the 
             `RADIUSS Shared CI <https://github.com/LLNL/radiuss-shared-ci>`_, 
             and the job implementation can be adapted. Please add a comment to
             describe the nature of the change.

```

rhornung67 replied:
done.

rhornung67 commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,257 @@
+  * `Spack <https://github.com/spack/spack>`_, a widely used
+    multi-platform package manager that builds and installs software stacks.
+  * `Uberenv <https://github.com/LLNL/uberenv>`_, a Python script
+    that helps to automate use of Spack and other tools for building third-party
```
Fixed.

rhornung67 commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,257 @@
+illustrated in the `RADIUSS Shared CI User Guide <https://radiuss-shared-ci.readthedocs.io/en/latest/sphinx/user_guide/index.html>`_. The guide also describes 
+how the framework works and how to set up a project to use it.
+
+In the next several sections, we describe files in the RAJA repo that are
+specific to the RAJA project.
```
Changed

rhornung67 commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,257 @@
+
+Briefly, these files play the following roles in our GitLab CI testing:
+
+  * The ``RAJA/.gitlab-ci.yml`` file defines general information that applies
+    to all GitLab CI test jobs run by the RAJA project.
```
changed.

rhornung67 commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,168 +12,257 @@
 
 In the following sections, we discuss how these files are used in the 
-steps in the RAJA Gitlab CI testing process summarized above.
+steps in the RAJA GitLab CI testing process summarized above.
```
changed
